### PR TITLE
Return null columns in result to match previous behaviour

### PR DESCRIPTION
**Pull request by wendigo**
https://github.com/prestosql/presto/pull/2860 broke `TestingPrestoClient` inside `addResults` function:

```
if (types.get() == null && statusInfo.getColumns() != null) {
     types.set(getTypes(statusInfo.getColumns()));
}
```

This PR restores old behaviour.

Fixes https://github.com/prestosql/presto/issues/4169

### CODE REVIEW

File: presto-main/src/main/java/io/prestosql/server/protocol/QueryResultRows.java [hunk 1]
```diff
@@ -90,8 +92,13 @@ public boolean isEmpty()
         return totalRows == 0;
     }
 
+    @Nullable
     public List<Column> getColumns()
     {
+        if (columns.isEmpty()) {
```
Comment (losipiuk) [inline presto-main/src/main/java/io/prestosql/server/protocol/QueryResultRows.java:98]: I do not know this part of code so I do not want to make final call. 
The question which comes to my mind is: are there legitimate cases when column list is empty and we want to distinguish it from null?

**Comment by findepi** [comment]
See https://github.com/prestosql/presto-python-client/blob/master/presto/dbapi.py#L196

**Comment by findepi** [comment]
Consider adding a test (maybe TestServer?)

Also, changing to `@Nullable` or -- better -- `Optinoal<List<Column>>` in `QueryResultRows` constructor would capture intent better.

Especially that -- purely in theory perhaps -- empty columns could be legit descrption of results (it's not today).

File: presto-main/src/main/java/io/prestosql/server/protocol/QueryResultRows.java (hunk 1 above)
Comment (wendigo) [reply presto-main/src/main/java/io/prestosql/server/protocol/QueryResultRows.java:98]: Initial query results does not contain column list at all. Previously null was returned, now it's empty array which breaks some clients (as null is part of the protocol)

**Comment by wendigo** [comment]
AC @findepi

File: presto-client/src/main/java/io/prestosql/client/QueryResults.java [hunk 2]
```diff
@@ -79,7 +80,7 @@ public QueryResults(
             URI infoUri,
             URI partialCancelUri,
             URI nextUri,
-            List<Column> columns,
+            Optional<List<Column>> columns,
```
Comment (findepi) [inline presto-client/src/main/java/io/prestosql/client/QueryResults.java:83]: Either revert the change here, or update `data` to be Optional as well

File: presto-main/src/main/java/io/prestosql/server/protocol/QueryResultRows.java [hunk 3]
```diff
@@ -82,15 +82,15 @@ private QueryResultRows(Session session, List<Type> types, List<Column> columns,
         this.currentPage = this.pages.pollFirst();
         this.supportsParametricDateTime = session.getClientCapabilities().contains(ClientCapabilities.PARAMETRIC_DATETIME.toString());
 
-        verify(this.types.size() == this.columns.size(), "columns and types sizes mismatch");
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/server/protocol/QueryResultRows.java:85]: should `types` be treated the same? ie also become Optional?

File: presto-main/src/main/java/io/prestosql/server/protocol/QueryResultRows.java [hunk 4]
```diff
@@ -107,7 +107,14 @@ public long getTotalRowsCount()
     public Optional<Long> getUpdateCount()
     {
         // We should have exactly single bigint value as an update count.
-        if (totalRows != 1 || columns.size() != 1 || !columns.get(0).getType().equals(StandardTypes.BIGINT)) {
+        if (totalRows != 1 || columns.isEmpty()) {
+            return Optional.empty();
+        }
+
+        verify(columns.isPresent(), "columns are not present");
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/server/protocol/QueryResultRows.java:114]: you can this right above, so that feels redundnat

File: presto-main/src/main/java/io/prestosql/server/protocol/QueryResultRows.java [hunk 5]
```diff
@@ -107,7 +107,14 @@ public long getTotalRowsCount()
...
+        List<Column> currentColumns = columns.get();
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/server/protocol/QueryResultRows.java:115]: `current` implies this changes over time, while column information does not.

you can use `columns` for the local var (yes, shadowing the field)

File: presto-main/src/main/java/io/prestosql/server/protocol/QueryResultRows.java [hunk 6]
```diff
@@ -150,7 +157,8 @@ public long getTotalRowsCount()
 
     private Optional<List<Object>> getRowValues()
     {
-        List<Object> row = new ArrayList<>(columns.size());
+        verify(columns.isPresent(), "columns are not present");
+        List<Object> row = new ArrayList<>(columns.get().size());
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/server/protocol/QueryResultRows.java:161]: add a ctor validation that presence of data implies presence of columns

File: presto-client/src/main/java/io/prestosql/client/QueryResults.java (hunk 2 above)
Comment (wendigo) [reply presto-client/src/main/java/io/prestosql/client/QueryResults.java:83]: I'll revert.

File: presto-main/src/main/java/io/prestosql/server/protocol/QueryResultRows.java (hunk 3 above)
Comment (wendigo) [reply presto-main/src/main/java/io/prestosql/server/protocol/QueryResultRows.java:85]: Done.

**Comment by wendigo** [comment]
AC @findepi. I've simplified checks and added couple of tests

**Comment by wendigo** [comment]
`2020-06-23T11:21:54.2603909Z io.prestosql.spi.PrestoException: The TCP/IP connection to the host localhost, port 32788 has failed. Error: "Connection refused (Connection refused). Verify the connection properties. Make sure that an instance of SQL Server is running on the host and accepting TCP/IP connections at the port. Make sure that TCP connections to the port are not blocked by a firewall.".`

File: presto-main/src/main/java/io/prestosql/server/protocol/QueryResultRows.java [hunk 7]
```diff
@@ -66,33 +64,27 @@
...
     public boolean isEmpty()
     {
         return totalRows == 0;
     }
 
-    public List<Column> getColumns()
+    public Optional<List<Column>> getColumns()
     {
-        return columns;
+        return columns.map(columns -> columns.stream().map(ColumnAndType::getColumn).collect(toUnmodifiableList()));
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/server/protocol/QueryResultRows.java:87]:
```suggestion
        return columns.map(columns -> columns.stream()
        .map(ColumnAndType::getColumn)
        .collect(toImmutableList()));
```

File: presto-main/src/main/java/io/prestosql/server/protocol/QueryResultRows.java [hunk 8]
```diff
@@ -150,10 +148,13 @@ public long getTotalRowsCount()
 
     private Optional<List<Object>> getRowValues()
     {
+        // types are present if data is present
+        List<ColumnAndType> columns = this.columns.get();
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/server/protocol/QueryResultRows.java:152]:
```suggestion
        List<ColumnAndType> columns = this.columns.orElseThrow();
```

(`get()` triggers an IDE warning)

File: presto-main/src/main/java/io/prestosql/server/protocol/QueryResultRows.java [hunk 9]
```diff
@@ -274,10 +270,57 @@ public QueryResultRows build()
...
+
+            verify(columns.size() == types.size(), "columns and types size mismatch");
+
+            ImmutableList.Builder<ColumnAndType> builder = ImmutableList.builder();
+
+            for (int i = 0; i < columns.size(); i++) {
+                builder.add(new ColumnAndType(columns.get(i), types.get(i)));
+            }
+
+            return Optional.of(builder.build());
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/server/protocol/QueryResultRows.java]:
```suggestion
            checkArgument(columns != null && types != null, "columns and types must be present at the same time");
            checkArgument(columns.size() == types.size(), "columns and types size mismatch");

            ImmutableList.Builder<ColumnAndType> builder = ImmutableList.builder();
            for (int i = 0; i < columns.size(); i++) {
                builder.add(new ColumnAndType(columns.get(i), types.get(i)));
            }
            return Optional.of(builder.build());
```

File: presto-main/src/main/java/io/prestosql/server/protocol/QueryResultRows.java [hunk 10]
```diff
@@ -274,10 +270,57 @@ public QueryResultRows build()
...
+        }
+    }
+
+    private static class ColumnAndType
+    {
+        private final Column column;
+        private final Type type;
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/server/protocol/QueryResultRows.java]: That's a bit unintuitive because `Column` also has type information (`String type` and `ClientTypeSignature typeSignature`)

cc @electrum ?

File: presto-main/src/main/java/io/prestosql/server/protocol/QueryResultRows.java [hunk 11]
```diff
@@ -66,33 +64,27 @@
...
-        this.types = requireNonNull(types, "types is null");
         this.columns = requireNonNull(columns, "columns is null");
         this.pages = new ArrayDeque<>(requireNonNull(pages, "pages is null"));
         this.exceptionConsumer = Optional.ofNullable(exceptionConsumer);
         this.totalRows = countRows(pages);
         this.currentPage = this.pages.pollFirst();
         this.supportsParametricDateTime = session.getClientCapabilities().contains(ClientCapabilities.PARAMETRIC_DATETIME.toString());
 
-        verify(this.types.size() == this.columns.size(), "columns and types sizes mismatch");
+        verify(totalRows == 0 || (totalRows > 0 && columns.isPresent()), "data present without columns and types");
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/server/protocol/QueryResultRows.java:77]: This is a new verification, it should be added as a separate commit.
(when splitting, make sure to split the tests accordingly too)

File: presto-main/src/main/java/io/prestosql/server/protocol/QueryResultRows.java (hunk 10 above)
Comment (wendigo) [reply presto-main/src/main/java/io/prestosql/server/protocol/QueryResultRows.java]: I followed logic of previous implementation. Columns are types are collected in `setQueryOutputInfo` method:

```
private synchronized void setQueryOutputInfo(QueryExecution.QueryOutputInfo outputInfo)
    {
        // if first callback, set column names
        if (columns == null) {
            List<String> columnNames = outputInfo.getColumnNames();
            List<Type> columnTypes = outputInfo.getColumnTypes();
            checkArgument(columnNames.size() == columnTypes.size(), "Column names and types size mismatch");

            ImmutableList.Builder<Column> list = ImmutableList.builder();
            for (int i = 0; i < columnNames.size(); i++) {
                list.add(createColumn(columnNames.get(i), columnTypes.get(i)));
            }
            columns = list.build();
            types = outputInfo.getColumnTypes();
        }

        for (URI outputLocation : outputInfo.getBufferLocations()) {
            exchangeClient.addLocation(outputLocation);
        }
        if (outputInfo.isNoMoreBufferLocations()) {
            exchangeClient.noMoreLocations();
        }
    }
```

**Comment by wendigo** [comment]
AC @findepi

File: presto-main/src/test/java/io/prestosql/server/TestServer.java [hunk 12]
```diff
@@ -127,6 +128,27 @@ public void testInvalidSessionError()
...
+                .map(JsonResponse::getValue)
+                .collect(toImmutableList());
+
+        QueryResults first = queryResults.get(0);
+        QueryResults last = queryResults.get(queryResults.size() - 1);
+
+        assertNull(first.getColumns());
+        assertEquals(first.getStats().getState(), "QUEUED");
+
+        assertNotNull(last.getColumns());
```
Comment (electrum) [inline presto-main/src/test/java/io/prestosql/server/TestServer.java:148]: Let's verify the contents as well

File: presto-main/src/test/java/io/prestosql/server/protocol/TestQueryResultRows.java [hunk 13]
```diff
@@ -65,7 +66,7 @@ public void shouldReturnSingleValue()
 
         assertThat((Iterable<? extends List<Object>>) rows).as("rows").isNotEmpty();
         assertThat(getAllValues(rows)).hasSize(1).containsOnly(ImmutableList.of(true));
-        assertThat(rows.getColumns()).containsOnly(column);
+        assertThat(rows.getColumns().get()).containsOnly(column);
```
Comment (electrum) [inline presto-main/src/test/java/io/prestosql/server/protocol/TestQueryResultRows.java:70]: Use `orElseThrow()` for all of these `get()` calls as that's the new preferred method when the value presence is not previously checked (so the expected behavior of throwing is clear).

File: presto-main/src/test/java/io/prestosql/server/protocol/TestQueryResultRows.java [hunk 14]
```diff
@@ -232,6 +233,59 @@ public void shouldHandleNullValues()
...
+                .withColumnsAndTypes(columns, types)
+                .build();
+    }
+
+    @Test(expectedExceptions = IllegalArgumentException.class, expectedExceptionsMessageRegExp = "columns and types must be present at the same time")
+    public void shouldThrowWhenColumnsAreNull()
+    {
+        List<Type> types = ImmutableList.of(IntegerType.INTEGER, BooleanType.BOOLEAN);
+
+        List<Page> pages = rowPagesBuilder(types)
```
Comment (electrum) [inline presto-main/src/test/java/io/prestosql/server/protocol/TestQueryResultRows.java:269]: Maybe extract a constant for this since it's shared across tests.

File: presto-main/src/main/java/io/prestosql/server/protocol/QueryResultRows.java [hunk 15]
```diff
@@ -178,10 +178,10 @@ else if (type instanceof TimestampWithTimeZoneType && !supportsParametricDateTim
         return Optional.of(unmodifiableList(row));
     }
 
-    private void propagateException(int row, int column, ColumnAndType columnAndType, Throwable cause)
+    private void propagateException(int row, ColumnAndType columnAndType, Throwable cause)
```
Comment (electrum) [inline presto-main/src/main/java/io/prestosql/server/protocol/QueryResultRows.java:186]: We could shorten this to `column` now

File: presto-main/src/main/java/io/prestosql/server/protocol/QueryResultRows.java [hunk 16]
```diff
@@ -280,6 +282,10 @@ public QueryResultRows build()
 
         private static Optional<List<ColumnAndType>> combine(List<Column> columns, List<Type> types)
```
Comment (electrum) [inline presto-main/src/main/java/io/prestosql/server/protocol/QueryResultRows.java:283]: We'd need to annotate these parameters with `@Nullable` as well. Instead of pushing nulls further, let's move the null checks into the caller.

File: presto-main/src/test/java/io/prestosql/server/TestServer.java (hunk 12 above)
Comment (electrum) [reply presto-main/src/test/java/io/prestosql/server/TestServer.java:152]: See comment. We should verify the contents here

File: presto-main/src/main/java/io/prestosql/server/protocol/QueryResultRows.java [hunk 17]
```diff
@@ -274,10 +275,54 @@ public QueryResultRows build()
...
                     session,
-                    types,
                     columns,
                     pages.build(),
                     exceptionConsumer);
         }
+
+        private static List<ColumnAndType> combine(List<Column> columns, List<Type> types)
+        {
+            checkArgument(columns != null && types != null, "columns and types must be present at the same time");
```
[the remark this answers is not in the document]
Comment (wendigo) [reply presto-main/src/main/java/io/prestosql/server/protocol/QueryResultRows.java:288]: I'll annotate

**Comment by findepi** [comment]
CI failed -- https://github.com/prestosql/presto/issues/2797
